### DISCUSSION TITLE
Override install_dir in command-line/automatically modify meson.build

### DISCUSSION THREAD

**Discussion by Stefan-Olt** [mesonbuild/meson#13719] [Q&A]
Hi,
is there any way to force meson with a command-line option to ignore the "install_dir" option in the meson.build file and install libraries/executables in the default location?
If not, is there a simple way to automatically edit a meson.build file and just delete that option? I tried with sed, but with regular expressions this is not possible reliable. I need that for an automated build/packaging system, so manual editing of the meson.build is not an option.
Best regards
Stefan

**Comment by eli-schwartz** [comment]
> I need that for an automated build/packaging system, so manual editing of the meson.build is not an option.

Why do you need that for an automated build/packaging system? At various times in my life I've developed and packaged software for multiple automated build/packaging systems and I have never before heard of this need.

If a project's meson.build file includes an install_dir doing something like e.g. `install_dir: get_option('libexecdir') / meson.project_name()` then I would expect the runtime usage of the software to simply abort on startup if it tries to run internal components and cannot do so because they are in the wrong location.

Can you show an example of a software package that needs the override you are asking about?

**Reply by Stefan-Olt** [reply]
It's about vapoursynth plugins, these are shared libraries (very rarely an executable). Some meson build scripts install them into the default library location, others detect the location of the vapoursynth library and install them into an subdirectory `vapoursynth` at that place (usually `/usr/local/lib/vapoursynth`). The second way is fine if you want to install them on your system. But I want the files to be installed to `{prefix}/lib`, so I can directly package them from there and do not install anything into the system.
Example of meson.build file: https://github.com/HomeOfVapourSynthEvolution/VapourSynth-AddGrain/blob/master/meson.build

**Reply by eli-schwartz** [reply]
> It's about vapoursynth plugins, these are shared libraries (very rarely an executable). Some meson build scripts install them into the default library location, others detect the location of the vapoursynth library and install them into an subdirectory `vapoursynth` at that place (usually `/usr/local/lib/vapoursynth`).

shared_module() is used for dlopen-style plugins. Usually, the way to install such a thing is to use `install_dir: get_option('libdir') / 'foobar'` where "foobar" is the name of the project that it's a plugin *for*, not the name of the plugin itself.

Here, too, I don't see the reason for overriding the install_dir. You want to package to `{prefix}/lib/vapoursynth` since that is the "plugin directory", and it is not meant to be linked to with `gcc foo-main.c -o foo-main -laddgrain`.

**Reply by Stefan-Olt** [reply]
Yes, that makes sense for installing the plugin in the system. But I do not want to install the plugin in the system, I want the shared library to be out into a specific directory so I can package it. Usually this is done by defining `--prefix`, so it will be put into `{prefix}/lib`, but install_dir in the meson.build file breaks `--prefix`.

This is not a standard unix-style package like .deb, as the install directory is not defined by the package, but by the package installer (a user cannot install to /usr/..., but he can still install the plugins to another directory), so only the file itself is packed, no directory structure.

I know this is a special use case, so is there an option like `--force-default-install-dir` or `--force-library-install-dir={prefix}/lib` and if not, is there a tool/method to edit a meson.build file so I can remove the install_dir option before running meson?

**Reply by eli-schwartz** [reply]
Can you just
```
meson install --destdir $PWD/temp_install/
```

And then cp around the ./temp_install/usr/lib/vapoursynth/libaddgrain.so to wherever you like?

> This is not a standard unix-style package like .deb, as the install directory is not defined by the package, but by the package installer (a user cannot install to /usr/..., but he can still install the plugins to another directory), so only the file itself is packed, no directory structure.

I've seen this sort of layout a bunch of times too -- in my experience, usually that would mean something like, the user installs all packages to $HOME/prefix-system/usr/{bin,include,lib,share} where --prefix=$HOME/prefix-system/usr

**Reply by Stefan-Olt** [reply]
But that still makes automation somewhat tricky, because it's not consistent (not all plugins use install_dir).
What I found is this: `meson rewrite kwargs delete target {target} install_dir foobar`. (if there is no random argument at the end meson will crash with an unhandled exception).
I have not found a way that meson will list the targets, but this seems to be good enough:
`sed -n -z "s/.*shared_module[ \t\n]*[(][ \t\n]*'\([A-Za-z0-9_]\+\)'.*/\1\n/p" meson.build`
